partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt

```diff
@@ -566,113 +561,153 @@ internal class PlanTyper(private val env: Env) {
         override fun visitRexOpPathIndex(node: Rex.Op.Path.Index, ctx: StaticType?): Rex {
             val root = visitRex(node.root, node.root.type)
             val key = visitRex(node.key, node.key.type)
-            val elementTypes = root.type.allTypes.map { type ->
-                if (type !is ListType && type !is SexpType) {
-                    return@map MISSING
+
+            // Check Index Type
+            if (!key.type.mayBeType<IntType>()) {
+                return ProblemGenerator.missingRex(
+                    rexOpPathIndex(root, key),
+                    ProblemGenerator.expressionAlwaysReturnsMissing("Collections must be indexed with integers, found ${key.type}")
+                )
+            }
+
+            // Get Element Type(s)
+            val elementTypes = root.type.allTypes.mapNotNull { type ->
+                when (type) {
+                    is ListType -> type.elementType
+                    is SexpType -> type.elementType
+                    else -> null
                 }
-                (type as CollectionType).elementType
-            }.toSet()
+            }
 
-            // TODO: For now we just log a single error.
-            //  Ideally we can log more detailed information such as key not integer, etc.
-            if (elementTypes.all { it is MissingType } || key.type !is IntType) {
+            // Check that root is not literal missing
+            if (root.isLiteralMissing()) {
                 return ProblemGenerator.missingRex(
                     rexOpPathIndex(root, key),
-                    ProblemGenerator.expressionAlwaysReturnsMissing("Path Navigation always returns MISSING")
+                    ProblemGenerator.expressionAlwaysReturnsMissing()
                 )
             }
-            val finalType = unionOf(elementTypes).flatten()
-            return rex(finalType.swallowAny(), rexOpPathIndex(root, key))
+
+            // Check that Root was LIST or SEXP by checking accumuated element types
+            if (elementTypes.isEmpty()) {
+                return ProblemGenerator.missingRex(
+                    rexOpPathIndex(root, key),
+                    ProblemGenerator.expressionAlwaysReturnsMissing("Only lists and s-expressions can be indexed with integers, found ${root.type}")
+                )
+            }
+            return rex(unionOf(elementTypes), rexOpPathIndex(root, key))
         }
 
+        private fun Rex.isLiteralMissing(): Boolean = this.op is Rex.Op.Lit && this.op.value.withoutAnnotations() == missingValue()
```

Comment:
Wondering if we could just do:

```suggestion
        private fun Rex.isLiteralMissing(): Boolean = this.op is Rex.Op.Lit && this.op.value is MissingValue
```

or are the annotations part of this check?